Relay runtime: add a new 'log' interface

mrtnzlml opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/facebook/relay/blob/1894b01ccbcddd12ac022f1660af6f692216ebe4/packages/relay-runtime/store/RelayModernEnvironment.js#L67, https://github.com/facebook/relay/blob/1894b01ccbcddd12ac022f1660af6f692216ebe4/packages/relay-runtime/store/RelayStoreTypes.js#L398-L438
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

mrtnzlml commented:
I'd need a little help with the remaining tics. How do I test this change properly in my project? How should I test it here? And how does the versioning work?

I'd also like to remove the `loggerProvider` which no longer exists. How to do it properly? Can I just delete it without worries?

renanmav commented on types/relay-runtime/lib/store/RelayStoreTypes.d.ts:
```diff
@@ -395,6 +395,67 @@ export interface RecordSourceSelectorProxy<T = {}> extends RecordSourceProxy {
 
+interface LogEventQueryResourceFetch {
+    readonly name: 'queryresource.fetch';
+    readonly operation: unknown;
```
Why is this `unknown`? Check [this](https://github.com/facebook/relay/blob/master/packages/relay-runtime/store/RelayStoreTypes.js#L401) and [this](https://github.com/facebook/relay/blob/master/packages/relay-runtime/store/RelayStoreTypes.js#L117-L121).

alloy commented:
> How do I test this change properly in my project?

You should be able to link the package, although I often just paste it into `node_modules/@types/relay-runtime` 🤷‍♂ 

> How should I test it here?

Add a real snippet of code that you expect to work and pass type-checking to the test file.

> And how does the versioning work?

Unsure what you mean by this, can you elaborate?

> I'd also like to remove the `loggerProvider` which no longer exists. How to do it properly? Can I just delete it without worries?

As long as [the minimum version of the package](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/a463a39a68a707f9dbf54e3b643945456221d86c/types/relay-runtime/index.d.ts#L1) that these typings support is correct (which it currently is) then you can safely remove older type defs 👍

mrtnzlml commented:
I fixed the review comments and removed the old logger interface (see: https://github.com/facebook/relay/commit/de62d7e9f181a3591a4b0c310d0d332ab23172e0). Can you please have a look again?

alloy commented:
Looks great, thanks for updating! 👌
